pex/executables.py

```diff
@@ -0,0 +1,147 @@
+    pattern=None,  # type: Optional[bytes]
+    check_executable=True,  # type: bool
+    extra_check=None,  # type: Optional[Callable[[bytes, BinaryIO], bool]]
+):
+    # type: (...) -> bool
+    """Determines if the given path is a script.
+
+    A script is a file that starts with a shebang (#!...) line.
+
+    :param path: The path to check.
+    :param pattern: Optional pattern to match against the shebang line (excluding the leading #!
+                    and trailing \n).
+    :param check_executable: Check that the script is executable by the current user.
+    :param extra_check: Optional callable accepting the shebang line (excluding the leading #! and
+                        trailing \n) and a file opened for binary read pointing just after that
+                        line.
+    :return: True if the given path is a script.
+    """
+    if check_executable and not is_exe(path):
+        return False
+    with open(path, "rb") as fp:
+        if b"#!" != fp.read(2):
+            return False
+        if not pattern:
+            return True
+        shebang_suffix = fp.readline().rstrip()
+        if bool(re.match(pattern, shebang_suffix)):
+            return True
+        if extra_check:
+            return extra_check(shebang_suffix, fp)
+        return False
+
+
+def create_sh_python_redirector_shebang(sh_script_content):
+    # type: (str) -> Tuple[str, str]
+    """Create a shebang block for a Python file that uses /bin/sh to find an appropriate Python.
+
+    The script should be POSIX compliant sh and terminate on all execution paths with an
+    explicit exit or exec.
+
+    The returned shebang block will include the leading `#!` but will not include a trailing new
+    line character.
+
+    :param sh_script_content: A POSIX compliant sh script that always explicitly terminates.
+    :return: A shebang line and trailing block of text that can be combined for use as a shebang
+             header for a Python file.
+    """
+    # This trick relies on /bin/sh being ubiquitous and the concordance of:
+    #
+    # 1. Python: Has triple quoted strings plus allowance for free-floating string values in
+    #    python files.
+    # 2. sh: Any number of pairs of `'` evaluating away when followed immediately by a
+    #    command string (`''command` -> `command`).
+    # 3. sh: The `:` noop command which accepts and discards arbitrary args.
+    #    See: https://pubs.opengroup.org/onlinepubs/009604599/utilities/colon.html
+    # 4. sh: Lazy parsing allowing for invalid sh content immediately following an exit or exec
+    #        line.
+    #
+    # The end result is a file that is both a valid sh script with a short shebang and a valid
+    # Python program.
+    return "#!/bin/sh", (
+        dedent(
+            """\
+            '''': pshprs
+            {sh_script_content}
+            '''
+            """
+        )
+        .format(sh_script_content=sh_script_content.rstrip())
+        .strip()
+    )
```

Comment:
The only bit of real interest in this PR. An old trick, but formalized and improved with the `: pshprs` sh comment allowing for efficient `is_python_script` checks of these otherwise not apparently Python scripts.

This trick is what allows using the existing `--sh-boot` PEX zip header script wholesale as a header to the existing PEX `__main__.py` launcher in loose and packed PEX directories.